pkg/controller/master/virtualmachine/vm_volume_status_controller.go

```diff
@@ -0,0 +1,493 @@
+	vmClient  ctlkubevirtv1.VirtualMachineClient
+	vmCache   ctlkubevirtv1.VirtualMachineCache
+	pvcClient corectl.PersistentVolumeClaimClient
+	pvcCache  corectl.PersistentVolumeClaimCache
+}
+
+type volumeStatus string
+
+type volumeStatusAnnotation struct {
+	// name of volume
+	Name string `json:"name,omitempty"`
+	// abstracted status based on long-horn (LH) volume status
+	Status string `json:"status,omitempty"`
+}
+
+type volumeStatusAnnotationSlice []volumeStatusAnnotation
+
+func (v volumeStatus) String() string {
+	return string(v)
+}
+
+// try to update LH Volume status to PVC Status
+func (h *VMVolumeStatusController) OnLHVolumeChange(key string, lhVolume *longhornv1.Volume) (*longhornv1.Volume, error) {
+	if lhVolume == nil || lhVolume.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	// when volume is not connected to pvc, skip
+	pvcName := lhVolume.Status.KubernetesStatus.PVCName
+	namespace := lhVolume.Status.KubernetesStatus.Namespace
+	if pvcName == "" || namespace == "" {
+		return nil, nil
+	}
+
+	volStatus := volumeStatusOK
+	// before LH has a better solution, we can only hard code those conditions to deduce one volume is in none-OK state
+	// new status should also be generated here
+	if lhVolume.Status.State == longhornv1.VolumeStateDetached && lhVolume.Status.Robustness == longhornv1.VolumeRobustnessUnknown {
+		if c, ok := lhVolume.Status.Conditions[longhornv1.VolumeConditionTypeScheduled]; ok {
+			if c.Status == longhornv1.ConditionStatusFalse && c.Reason == longhornv1.VolumeConditionReasonReplicaSchedulingFailure {
+				// assume insufficient storage
+				volStatus = volumeStatusInsufficientStorage
+			}
+		}
+	}
+	return nil, h.updateLHVolumeStatusToPVC(lhVolume.Name, namespace, pvcName, volStatus)
+}
+
+// try to remove volume status from PVC Status if it is there
+func (h *VMVolumeStatusController) OnLHVolumeRemove(key string, lhVolume *longhornv1.Volume) (*longhornv1.Volume, error) {
+	if lhVolume == nil {
+		return nil, nil
+	}
+
+	pvcName := lhVolume.Status.KubernetesStatus.PVCName
+	namespace := lhVolume.Status.KubernetesStatus.Namespace
+	if pvcName == "" || namespace == "" {
+		return nil, nil
+	}
+
+	return nil, h.removeLHVolumeStatusFromPVC(lhVolume.Name, namespace, pvcName)
+}
+
+// LH volume is changed, update the possible volume status to PVC
+func (h *VMVolumeStatusController) updateLHVolumeStatusToPVC(volName string, pvcNamespace string, pvcName string, status volumeStatus) error {
+	var err error
+	var pvc *corev1.PersistentVolumeClaim
+	// pvc may be deleted, but LH volume is still there, in such case, give up updating
+	if pvc, err = h.getPVC(pvcNamespace, pvcName); err != nil {
+		if !errors.IsNotFound(err) {
+			logrus.Infof("fail to get pvc:(%s/%s), give up to update LH volume:%s status:%s, error:%s", pvcNamespace, pvcName, volName, status.String(), fmt.Errorf("%w", err))
+		}
+		return nil
+	}
+
+	if status == volumeStatusOK {
+		if _, ok := pvc.Annotations[util.AnnotationVolumeStatus]; ok {
```

Comment:
Candidate is putting` "volStatus, ok := pvc.Annotations[util.AnnotationVolumeStatus]"`  to L104, then in case "volumeStatusOK " go may complain `volStatus defined but not used` . 
Keep unchanged will not hurt too much.